### ISSUE TITLE
[Golang]: SSTI Method Confusion

### ISSUE THREAD

**Issue by aydinnyunus** [github/securitylab#812]
### Query PR

https://github.com/github/codeql/pull/15193

### Language

GoLang

### CVE(s) ID list

- [CVE-2023-2259](https://nvd.nist.gov/vuln/detail/CVE-2023-2259)  
- [CVE-2023-2017](https://nvd.nist.gov/vuln/detail/CVE-2023-2017) 
- [CVE-2022-47896](https://nvd.nist.gov/vuln/detail/CVE-2022-4789)  

### CWE

CWE-1336 - Improper Neutralization of Special Elements Used in a Template Engine

### Overview
This pull request introduces a new CodeQL query specifically designed to detect Server-Side Template Injection (SSTI) vulnerabilities in Go applications. SSTI in Go can occur when user input is dynamically embedded into templates, leading to potential code execution or other security breaches. This query focuses on identifying patterns in Go code that could be exploited through SSTI.

### Vulnerability

This function essentially takes in a string parameter called "file", which is the opened and read. This can be easily verified, although a slight modification is required to be made to the script, specifically, execute the template by taking in the Context data type, as File is a method of Context. In my testing environment, c is already type Context, so we can just pass the variable c to the execute function.

````
func (c *context) File(file string) (err error) {
	f, err := os.Open(file)
	if err != nil {
		return NotFoundHandler(c)
	}
	defer f.Close()

	fi, _ := f.Stat()
	if fi.IsDir() {
		file = filepath.Join(file, indexPage)
		f, err = os.Open(file)
		if err != nil {
			return NotFoundHandler(c)
		}
		defer f.Close()
		if fi, err = f.Stat(); err != nil {
			return
		}
	}
	http.ServeContent(c.Response(), c.Request(), fi.Name(), fi.ModTime(), f)
	return
}
````

This means we can succesfully abuse the gadget to read a local file. This similar tactic can be abused in many situations, as long as the data parameter is the same type as the one that the method uses, you can invoke said method with custom parameters.

![image](https://github.com/github/securitylab/assets/52822869/972a74d4-b794-4b02-b9df-8a02943957cc)

### Are you planning to discuss this vulnerability submission publicly? (Blog Post, social networks, etc).

- [X] Yes
- [ ] No

**Comment by sylwia-budzynska** [comment]
@aydinnyunus Could you provide a CodeQL database for one of the CVEs? Here I mean a CodeQL database not of the code examples that you provide, but a CodeQL database of one of the CVEs. For example, of CVE-2023-2259 (so a database of alfio-event/alf.io prior to 2.0-M4-2304 with the vulnerability present in the code). Your query should be able to find the CVE in that database as per [rule 4](https://securitylab.github.com/bounties/) in our bug bounty rules.

Another thing - your query currently does not produce any results for any of our MRVA lists. It could be that 1) it is not a widespread issue, 2) or that the go apps in our lists do use the go templates, 3) or that there is an issue with the CodeQL models in your query. It's also requirement as part of the [rule 4](https://securitylab.github.com/bounties/) `To be considered, your query must find at least one CVE that was not previously found by an existing query, in a released version (older releases are also permitted) of an open source project that is actually used (no demo, training, vulnerable on purpose). Submissions without at least one result won't be considered. `.

Do you have any MRVA lists with projects, that would have true positive results, that you can share here or privately?
Is there anything you could do to improve the modeling in your query (for example, add new CodeQL models)?

**Comment by aydinnyunus** [comment]
@sylwia-budzynska I did not find direct CVE for the provided query but the root cause of this vulnerability relies on SSTI Vulnerability on Template Engines so I attached the other CVEs.  This vulnerability exists and have many research posts so I create the CodeQL query. I did not try with all vulnerable codes because they are vulnerable on purpose.

- https://www.onsecurity.io/blog/go-ssti-method-research/
- https://dev.to/pirateducky/ssti-method-confusion-in-go-517p
- https://chinnidiwakar.gitbook.io/githubimport/pentesting-web/ssti-server-side-template-injection#method-confusion-in-go
